### PR TITLE
Fix ONNX dynamic-shape RNN capture

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184872

ONNX export with dynamic sequence lengths was still using the default RNN decompositions during torch.export capture. The default LSTM/GRU decompositions iterate over the precomputed sequence tensor in Python, which materializes the time dimension and forces a static guard, so a user-specified dynamic time dimension conflicts with tracing.

Use the existing while_loop-based LSTM and GRU decompositions when ONNX capture receives dynamic_shapes, so the sequence loop stays in the graph instead of being expanded by Python iteration. Flattened RNN weights can share storage, and higher-order op tracing rejects aliased closure inputs, so clone only the read-only recurrent weights that overlap before closing over them in the while_loop.

I considered limiting the ONNX patch to LSTM only, but GRU has the same Python sequence iteration pattern and already has a matching while_loop decomposition, so registering both keeps the RNN dynamic-shape behavior consistent.

Fixes #174272

Generated by my agent

Test Plan:
- python test/onnx/exporter/test_api.py -k test_dynamic_shape_bilstm_batch_first
- python test/export/test_export.py -k test_dynamic_lstm_with_aliased_flat_weights
- lintrunner -a
- git diff --check
- git diff --cached --check

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by justinchuby — APPROVED** [review]
Thanks!

File: test/onnx/exporter/test_api.py [hunk 1]
```diff
@@ -151,6 +151,27 @@ def forward(self, a, x):
 
         self.assertEqual(len(onnx_program.model.graph.inputs), 1)
 
+    def test_dynamic_shape_bilstm_batch_first(self):
+        class BiLSTM(torch.nn.Module):
```
Comment (justinchuby) [inline test/onnx/exporter/test_api.py]: Nit: these tests will fit better in the small model e2e file

Comment (jansel) [reply test/onnx/exporter/test_api.py]: My agent says I moved the dynamic-shape BiLSTM coverage from test_api.py into test_small_models_e2e.py, keeping the dynamic axis assertion and alternate sequence-length ONNXProgram check there.

**Review by zou3519 — APPROVED** [review]
